### PR TITLE
Improve IPVS Module loader logic

### LINKED ISSUE #77493: kube_proxy running under a static kernel reports false-positive module load failures
<Reopened this valid and in-demand bug after the auto-bot expired it>

/kind bug
/sig network
/area kube-proxy

**What happened**:
Building a fully static Linux kernel with  kube_proxy's desired IP_VS (Virtual Server) modules linked statically:

CONFIG_IP_VS=y
CONFIG_IP_VS_RR=y
CONFIG_IP_VS_WRR=y
CONFIG_IP_VS_SH=y
CONFIG_NF_CONNTRACK_IPV4=y

Causes kube_proxy to believe that these features don't exist because it can't find modules for them:

```
W0923 18:20:43.979259       1 proxier.go:469] Failed to load kernel module ip_vs with modprobe. You can ignore this message when kube-proxy is running inside container without mounting /lib/modules
W0923 18:20:43.979848       1 proxier.go:469] Failed to load kernel module ip_vs_rr with modprobe. You can ignore this message when kube-proxy is running inside container without mounting /lib/modules
W0923 18:20:43.980415       1 proxier.go:469] Failed to load kernel module ip_vs_wrr with modprobe. You can ignore this message when kube-proxy is running inside container without mounting /lib/modules
W0923 18:20:43.981481       1 proxier.go:469] Failed to load kernel module ip_vs_sh with modprobe. You can ignore this message when kube-proxy is running inside container without mounting /lib/modules
W0923 18:20:43.982052       1 proxier.go:469] Failed to load kernel module nf_conntrack_ipv4 with modprobe. You can ignore this message when kube-proxy is running inside container without mounting /lib/modules

```

**What you expected to happen**:

If Kube_proxy's desired kernel features are linked statically in the kernel, kube_proxy should carry on and not attempt and throw error messages about modules not being found.

Under the hood, kube_proxy should simply check for the existence of the desired kernel features in the /proc tree (ie: `/proc/sys/net/ipv4/vs`, and others) **before** attempting to look for and load modules. If, after module loads fails, then throw the error messages and switch to iptables.

**How to reproduce it (as minimally and precisely as possible)**:

1. Download and extract the current stable kernel source version (today it's 4.18.9) from kernel.org
2. Copy your existing running `/boot/config-old_kernel_version` into the source tree as `.config` 
3. Run `make oldconfig` and answer any questions
4. Run ` make localyesconfig` to switch to a fully static kernel
4. Modify `.config` and ensure the following exist, and are not `=m`:
    ```
    CONFIG_IP_VS=y
    CONFIG_IP_VS_RR=y
    CONFIG_IP_VS_WRR=y
    CONFIG_IP_VS_SH=y
    CONFIG_NF_CONNTRACK_IPV4=y
    ```
5. Run `make menuconfig` and deselect  `[ ] Enable loadable module support` (box is empty)
5. `make -j$(nproc)`
7. `cp arch/x86/boot/bzImage /boot/vmlinuz-4.18.9-kube_proxy_mods_static`
8. `cp .config /boot/config-4.18.9-kube_proxy_mods_static`
9. Regenerate your boot loader's config (`grub2-mkconfig -o /boot/grub2/grub.cfg`) or (`grub-mkconfig -o /boot/grub/grub.cfg`) depending on your distribution.
10. Reboot and select your new kernel with the kube_proxy modules linked statically.
11. Start kube_proxy and watch it throw the above errors about not finding the modules.

**Anything else we need to know?**:

Fully static kernels will not generate `/lib/modules/$kerne_version/modules.builtin` so inspecting that file will fail - as well in the kube_proxy source. The gold-standard is to simply check for the features in the active`/proc/...` tree first, and consider those as "built-in".   

The next best options is checking `/boot/config-$(uname -r)` for `CONFIG_<FEATURE>=y` statements, which tells you if they're built-in.  Note that `/boot/config-<version>` files are almost universally installed by OS maintainers.

And finally, you could try checking `/proc/config.gz`, which is a compressed version of the externally held .config text file, in a similar manner noted above, however it's up to the maintainer to build that feature into the kernel (so it might not exist). 

For reference, here's the list of modules kube_proxy wants: https://github.com/kubernetes/kubernetes/blob/master/pkg/proxy/ipvs/proxier.go#L161

Here we see kube_proxy's errant module-centric code that needs to be made compatible w/ static kernels:
https://github.com/kubernetes/kubernetes/blob/master/pkg/proxy/ipvs/proxier.go#L433

**Environment**:
- Kubernetes version (use `kubectl version`): 1.11.3
- Cloud provider or hardware configuration: On premise
- OS (e.g. from /etc/os-release): Leap 15.0
- Kernel (e.g. `uname -a`): 4.18.9 statically linked
- Install tools: gcc 8.1
- Others:

<reopened as it appears still unfixed and in-demand>  See how far the team got on it before it was auto-closed here: https://github.com/kubernetes/kubernetes/issues/69006%3E

### PR DESCRIPTION
**Pull request by rikatz**
Signed-off-by: Ricardo Pchevuzinske Katz <ricardo.katz@serpro.gov.br>

**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**: Improve kube-proxy IPVS Module loader logic and prepare for PR #82223 to check static compiled kernel

**Which issue(s) this PR fixes**:
Fixes #77493 (partially)

**Special notes for your reviewer**: This PR was opened as a request to split the logic improvement of PR #82223

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

### CODE REVIEW

File: pkg/proxy/ipvs/proxier.go [hunk 1]
```diff
@@ -578,33 +578,34 @@ func (handle *LinuxKernelHandler) GetModules() ([]string, error) {
 	}
 	ipvsModules := utilipvs.GetRequiredIPVSModules(kernelVersion)
 
+	var bmods []string
+
+	// Find out loaded kernel modules. If this is a full static kernel it will thrown an error
+	modulesFile, err := os.Open("/proc/modules")
+	if err != nil {
+		klog.Warningf("Failed to read file /proc/modules with error %v.  Kube-proxy doesn't support fully statically compiled kernel yet", err)
```
Comment (neolit123) [inline pkg/proxy/ipvs/proxier.go]: extra space before `Kube-proxy`

Comment (rikatz) [reply pkg/proxy/ipvs/proxier.go]: Corrected

File: pkg/proxy/ipvs/proxier.go [hunk 2]
```diff
@@ -578,33 +578,34 @@ func (handle *LinuxKernelHandler) GetModules() ([]string, error) {
 	}
 	ipvsModules := utilipvs.GetRequiredIPVSModules(kernelVersion)
 
+	var bmods []string
+
+	// Find out loaded kernel modules. If this is a full static kernel it will thrown an error
+	modulesFile, err := os.Open("/proc/modules")
```
Comment (andrewsykim) [inline pkg/proxy/ipvs/proxier.go]: Prefer to keep this check in the original location (line 593), close to where the output (`modulesFile`) is consumed.

Comment (rikatz) [reply pkg/proxy/ipvs/proxier.go]: @andrewsykim actually I consume the err here to check if this is not a static kernel. That's why I've changed the location

Comment (andrewsykim) [reply pkg/proxy/ipvs/proxier.go]: Gotcha, can we move 
```
	mods, err := getFirstColumn(modulesFile)
	if err != nil {
		return nil, fmt.Errorf("failed to find loaded kernel modules: %v", err)
	}
```

up here as well then?

Comment (rikatz) [reply pkg/proxy/ipvs/proxier.go]: For sure! Just did it

File: pkg/proxy/ipvs/proxier.go [hunk 3]
```diff
@@ -578,30 +578,12 @@ func (handle *LinuxKernelHandler) GetModules() ([]string, error) {
...
-			klog.Warningf("Failed to load kernel module %v with modprobe. "+
-				"You can ignore this message when kube-proxy is running inside container without mounting /lib/modules", kmod)
-		}
-	}
-
-	// Find out loaded kernel modules
+	// Find out loaded kernel modules. If this is a full static kernel it will thrown an error
 	modulesFile, err := os.Open("/proc/modules")
 	if err != nil {
+		klog.Warningf("Failed to read file /proc/modules with error %v. Kube-proxy doesn't support fully statically compiled kernel yet", err)
```
Comment (andrewsykim) [inline pkg/proxy/ipvs/proxier.go]: My knowledge in this area is lacking a bit. Wondering if "fully statically compiled kernel" is the right term here -- wondering if something like this is bit clearer to users:
```
Kube-proxy requires loadable modules enabled in the kernel
```

Comment (rikatz) [reply pkg/proxy/ipvs/proxier.go]: Done :)

**Comment by andrewsykim** [comment]
/approve
/lgtm

Thanks for iterating @rikatz

**Comment by rikatz** [comment]
/retest
Flaking tests :(

**Pull request merged**